nominal/core/stream.py

```diff
@@ -19,31 +21,38 @@ class BatchItem:
     channel_name: str
     timestamp: str | datetime | IntegralNanosecondsUTC
     value: float
-    tags: Dict[str, str] | None = None
+    tags: dict[str, str] | None = None
+
 
+NominalWriteStream: TypeAlias = "WriteStream"
 
-class NominalWriteStream:
+
+class WriteStream:
     def __init__(
         self,
         process_batch: Callable[[Sequence[BatchItem]], None],
         batch_size: int = 10,
-        max_wait_sec: int = 5,
+        max_wait: IntegralSecondsDuration = 5,
         max_workers: int | None = None,
     ):
         """Create the stream."""
         self._process_batch = process_batch
         self.batch_size = batch_size
-        self.max_wait_sec = max_wait_sec
-        self._executor = concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)
+        self.max_wait = max_wait
+        self.max_workers = max_workers
         self._batch: list[BatchItem] = []
         self._batch_lock = threading.Lock()
         self._last_batch_time = time.time()
         self._running = True
+        self._max_wait_event = threading.Event()
+        self._pending_jobs = threading.BoundedSemaphore(3)
 
+    def start(self) -> None:
+        self._executor = concurrent.futures.ThreadPoolExecutor(max_workers=self.max_workers)
         self._timeout_thread = threading.Thread(target=self._process_timeout_batches, daemon=True)
         self._timeout_thread.start()
 
-    def __enter__(self) -> "NominalWriteStream":
+    def __enter__(self) -> "WriteStream":
```

Comment:
we have `from __future__ import annotations` so we don't need to stringify the type here